### PR TITLE
part of the infrastructure for public & private dependencies in the resolver

### PR DESCRIPTION
**Pull request by Eh2406**
This is part of my work on public & private dependencies in the resolver from #6129. As discussed there the proptest fuzzers are happy to find exponential blow up with all the back jumping strategies I have tried. So this PR does not have a back jumping strategie nor does it have the proptest fuzzers generating public dependencies. These will both need to change for the feature to stabilize. In the meantime it gives the correct results on the cases it can handle.

With https://github.com/rust-lang/rust/pull/57586 landed there is a lot of work to do on Cargos front end. Adding a UI for this, passing the relevant things to rustc, passing it to crates.io, passing it to cargo-metadata. This is good enough to allow that work to proceed.

### CODE REVIEW

**Comment by Eh2406** [comment]
Push a fix for my braking every test.

File: src/cargo/core/resolver/context.rs [hunk 1]
```diff
@@ -26,6 +26,12 @@ pub struct Context {
     pub activations: Activations,
     pub resolve_features: im_rc::HashMap<PackageId, Rc<HashSet<InternedString>>>,
     pub links: im_rc::HashMap<InternedString, PackageId>,
+    pub public_dependency:
+        im_rc::HashMap<PackageId, im_rc::HashMap<InternedString, (PackageId, bool)>>,
```
Comment (alexcrichton) [inline src/cargo/core/resolver/context.rs]: We should probably already have this for all the other fields but could a comment be added here for this new field about what this is a map to/from?

File: src/cargo/core/resolver/context.rs [hunk 2]
```diff
@@ -26,6 +26,12 @@ pub struct Context {
...
+
+    // This is somewhat redundant with the `resolve_graph` that stores the same data,
+    //   but for querying in the opposite order.
+    pub parents: Graph<PackageId, Rc<Vec<Dependency>>>,
```
Comment (alexcrichton) [inline src/cargo/core/resolver/context.rs]: In our past debugging we've found that cloning a `Context` happens so often that an expensive clone means a slower resolver, so I was curious about this! 

I think we explicitly removed `Graph` from this data structure earlier in favor of `RcList` below, so was curious on your thoughts on this.

I haven't read the whole patch yet though, so I'll likely find what happens soon.

File: src/cargo/core/resolver/mod.rs [hunk 3]
```diff
@@ -591,6 +596,44 @@ fn activate(
...
+                .link(candidate.summary.package_id(), parent.package_id()),
+        )
+        .push(dep.clone());
+        let cs: Vec<PackageId> = cx
+            .public_dependency
+            .get(&candidate.summary.package_id())
+            .iter()
+            .flat_map(|x| x.values())
+            .filter_map(|x| if x.1 { Some(&x.0) } else { None })
+            .chain(Some(candidate.summary.package_id()).iter())
```
Comment (alexcrichton) [inline src/cargo/core/resolver/mod.rs]: The trailing `.iter()` here I think can be removed

File: src/cargo/core/resolver/mod.rs [hunk 4]
```diff
@@ -591,6 +596,44 @@ fn activate(
             candidate.summary.package_id(),
             dep.clone(),
         ));
+        Rc::make_mut(
```
Comment (alexcrichton) [inline src/cargo/core/resolver/mod.rs]: Could a comment be added here to explain what's going on?

File: src/cargo/core/resolver/mod.rs [hunk 5]
```diff
@@ -725,6 +769,30 @@ impl RemainingCandidates {
                     continue;
                 }
             }
+            for &t in cx
```
Comment (alexcrichton) [inline src/cargo/core/resolver/mod.rs]: Could a comment be added here abot what candidates this is rejecting?

File: src/cargo/core/resolver/mod.rs [hunk 6]
```diff
@@ -799,6 +868,9 @@ fn find_candidate(
         // make any progress. As a result if we hit this condition we can
         // completely skip this backtrack frame and move on to the next.
         if !backtracked
+            && !conflicting_activations
+                .values()
+                .any(|c| *c == ConflictReason::PublicDependency)
```
Comment (alexcrichton) [inline src/cargo/core/resolver/mod.rs]: Could the comment be updated to account for this new clause?

File: src/cargo/core/resolver/types.rs [hunk 7]
```diff
@@ -420,6 +425,8 @@ impl ConflictReason {
     }
 }
 
+pub type Conflict = BTreeMap<PackageId, ConflictReason>;
```
Comment (alexcrichton) [inline src/cargo/core/resolver/types.rs]: s/Conflict/ConflictMap/ perhaps?

File: src/cargo/util/graph.rs [hunk 8]
```diff
@@ -1,35 +1,36 @@
...
+use std::collections::BTreeSet;
 use std::fmt;
-use std::hash::Hash;
 
-pub struct Graph<N, E> {
-    nodes: HashMap<N, HashMap<N, E>>,
+use im_rc;
+
+pub struct Graph<N: Clone, E: Clone> {
+    nodes: im_rc::OrdMap<N, im_rc::OrdMap<N, E>>,
```
Comment (alexcrichton) [inline src/cargo/util/graph.rs]: aha, I see how this would make things a bit speedier on the cloning half of things

Comment (alexcrichton) [reply src/cargo/util/graph.rs]: (maybe `RcList` is now obsoleted?)

**Comment by Eh2406** [comment]
I added comments, I think it addresses your concerns.

The test point out that the RFC did not discuss how this interacts with the other things cargo can do. I don't have a mental model of how this should work.

**Comment by Eh2406** [comment]
So the specific questions are:

How does this interact with renamed dependencies?
--------
There are lots of juicy corner cases here. On one had being able to depend on 2 different versions of the same library is one of the points of renamed dependencies. Fore example so you can publicly derive `serde1` and `serde2` traits on the same type. On the other hand making sure you and your dependencies agree on the same version of `serde` is the hole point of public dependencies.

This makes my head hurt, is there something simple I am missing? If not, where is the appropriate place to discuss fundamental problems with an accepted RFC?

How does this interact with `'cfg({})'`ing in different versions?
----------
Conceptually this one is very simple. Instead of asking "is there a package that can see two different versions of a dependency" we need to ask "is there a set of arguments so that there is a package that can see two different versions of a dependency". But the complexity is in the details. If we are not careful this is a SAT problem that needs to be solved for each tick. Where is the existing `'cfg({})'` logic handled?

Is there some minimal subset of this feature that we can get to merge while we work the issues out? Like how this PR does not include backtracking nor error messages.

**Comment by Eh2406** [comment]
I got test to pass by just making deps that have a cfg or a rename invisible. This is not a sablizable solution, but it does get the test to pass.

File: src/cargo/core/resolver/mod.rs [hunk 9]
```diff
@@ -585,35 +590,80 @@ fn activate(
...
+            cx.parents.link(candidate_pid, parent_pid),
+        )
+        // and associate dep with that edge
+        .push(dep.clone());
+        // For now the interaction of public dependency conflicts with renamed and
+        // 'cfg({})' dependencies is not clear. So we disable public checks if the
+        // other functionality is used.
+        // TODO: this disable is not a long term solution.
+        //       This needs to be removed before public dependencies are stabilized!
+        if dep.platform().is_none() && dep.explicit_name_in_toml().is_none() {
```
Comment (alexcrichton) [inline src/cargo/core/resolver/mod.rs]: I think the behavior that we want is to probably just delete this `if` clause and unconditionally run the below?

**Comment by alexcrichton** [comment]
Do you think that we'll need to have fast paths for no public dependencies to preserve the current performance of the resolver? Or do the new code paths largely come out in the wash in profiling?

File: src/cargo/core/resolver/mod.rs (hunk 9 above)
Comment (Eh2406) [reply src/cargo/core/resolver/mod.rs]: Yes, we would like to delete this if statement and have the code below deal with the special handling of these corner cases. At the beginning of this PR we did not have this if statement and it broke the existing tests. Until we figure out how the corner cases are supposed to be handled, this if statement lets other people make progress.

**Comment by Eh2406** [comment]
Sorry for the delay in replying, real life is being very difficult at the moment.

From what I have seen, this is not a hotspot. I have mostly been investigating cases of exponential blowout caused by this feature, and even there this inefficient code is not hot. I think with the use of Im-rs the general performance regression will will not be significant in absolute terms. That being said if you would like to check servo and x.py, it can't hurt.

**Comment by Eh2406** [comment]
Life has been kicking me this week, and will be for the foreseeable future. I found 2 mins to rebase this. I think let's get this landed and fix the problems in follow up PRs.

File: src/cargo/core/resolver/mod.rs (hunk 9 above)
Comment (alexcrichton) [reply src/cargo/core/resolver/mod.rs]: Er sorry it's taking me awhile to get back to this, but do you have a corner case in mind that this is needed for? I can't actually think of a case where the platform and/or renaming would cause issues

File: src/cargo/core/resolver/mod.rs [hunk 10]
```diff
@@ -585,35 +590,80 @@ fn activate(
...
+            let cs: Vec<PackageId> = cx
+                .public_dependency
+                .get(&candidate_pid)
+                .iter()
+                .flat_map(|x| x.values())
+                .filter_map(|x| if x.1 { Some(&x.0) } else { None })
+                .chain(&Some(candidate_pid))
+                .cloned()
+                .collect();
+            for c in cs {
```
Comment (alexcrichton) [inline src/cargo/core/resolver/mod.rs]: Could the documentation for this loop and the `cs` variable above be expanded to include what they're doing?

**Comment by alexcrichton** [comment]
Ok I think if it's not showing up anywhere in profiles yet we should be good to add this, and we can always use bug reports on nightly to help guide us! I think with a few more comments (listed above) and the removal of the cfg/naming branches this should be good to land

File: src/cargo/core/resolver/mod.rs (hunk 9 above)
Comment (Eh2406) [reply src/cargo/core/resolver/mod.rs]: It is easy to find the examples, just look at the test that do not pass without it.
- renaming: [this test](https://github.com/rust-lang/cargo/blob/d75d1fbe6b43e59d0a04e73659e411855c2633fc/tests/testsuite/rename_deps.rs#L7) means that `foo` can see `bar 0.1.0` and `bar 0.2.0`. `foo` can get `expected bar::Thing but got bar::Thing` errores so is not permitted by public & private dependencies.
- platform: [this test](https://github.com/rust-lang/cargo/blob/1e768f712dfac8c093d46910a7582a3aec0e03de/tests/testsuite/cfg.rs) means that `foo` can see `bar 0.1.0` and `bar 0.2.0` depending on platform. This should not be a problem, but requires a deep level of disjointness analysis about platforms to be added to make it work.

Comment (alexcrichton) [reply src/cargo/core/resolver/mod.rs]: For renaming though we want to give an error in that case, so renaming doesn't affect public/private, right? For platforms it's true that in theory we want to take the exhaustive list of any platforms and if none of them actually activate both dependencies we don't give an error, but it seems that for all practical purposes we'd want to give an error assuming all dependencies can be activated on at least one platform.

File: src/cargo/core/resolver/mod.rs (hunk 10 above)
Comment (Eh2406) [reply src/cargo/core/resolver/mod.rs]: Added documentation for the `cs` variable. Now that I am typing a response I see that you wanted an explanation of the entire loop. BRB.

File: src/cargo/core/resolver/mod.rs (hunk 9 above)
Comment (Eh2406) [reply src/cargo/core/resolver/mod.rs]: That is a reasonable position. (although different from what you suggested at https://github.com/rust-lang/cargo/issues/6129#issuecomment-465189082) It would mean that this existing stable behavior should not be allowed. Such a decision would require more discussion. I thought that we could have that discussion in a follow up PR if that is the behavior we decide on. 

This PR is the minimum functionality to unlock further work. I thought it would be more in kind for this PR to adopt the behavior of treating `cfg` and `rename` as having the "escape hatch" you suggested.

File: src/cargo/core/resolver/mod.rs (hunk 10 above)
Comment (Eh2406) [reply src/cargo/core/resolver/mod.rs]: Comments updated more fully. Eventually this should probably be a dedicated data structure that keeps the related functionality together, and removes some of the duplication.

File: src/cargo/core/resolver/mod.rs [hunk 11]
```diff
@@ -487,6 +527,7 @@ fn activate_deps_loop(
                                 &parent,
                                 backtracked,
                                 &conflicting_activations,
+                                check_public_visible_dependencies,
```
Comment (alexcrichton) [inline src/cargo/core/resolver/mod.rs]: Could this perhaps be stored in a `Context` or somewhere to avoid being passed around?

**Comment by alexcrichton** [comment]
r=me with just one small stylistic nit about passing variables around

File: src/cargo/core/resolver/mod.rs (hunk 11 above)
Comment (Eh2406) [reply src/cargo/core/resolver/mod.rs]: It could, and if you want to do it will be. Generally, Context storage things that change as resolution precedes. This will always have the same value. So I didn't see the need to store it repeatedly on the heap. Of course it's one bool, so it doesn't matter very much.

Would you like me to change it?

Comment (alexcrichton) [reply src/cargo/core/resolver/mod.rs]: Yeah it's not the greatest spot for it, but I figure it's a bit better than passing and threading around to lots of places

Comment (Eh2406) [reply src/cargo/core/resolver/mod.rs]: Does that look better?

**Comment by alexcrichton** [comment]
@bors: r+

Looks great to me!
